_posts/2021-08-16-connector-table-sql-api-part2.md

```diff
@@ -0,0 +1,434 @@
+    public DynamicTableSource copy() {
+        return new ImapTableSource(options, columnNames);
+    }
+
+    // …
+}
+```
+
+Finally, in the `ImapTableSourceFactory` class, you need to create a `ConfigOption<Type>Name` for the hostname, port number, username, and password.  Then you need to report them to Flink. Since all of the current options are mandatory, you can add them to the `requiredOptions()` method in order to do this. 
+
+```java
+public class ImapTableSourceFactory implements DynamicTableSourceFactory {
+
+    public static final ConfigOption<String> HOST = ConfigOptions.key("host").stringType().noDefaultValue();
+    public static final ConfigOption<Integer> PORT = ConfigOptions.key("port").intType().noDefaultValue();
+    public static final ConfigOption<String> USER = ConfigOptions.key("user").stringType().noDefaultValue();
+    public static final ConfigOption<String> PASSWORD = ConfigOptions.key("password").stringType().noDefaultValue();
+
+    // …
+
+    @Override
+    public Set<ConfigOption<?>> requiredOptions() {
+        final Set<ConfigOption<?>> options = new HashSet<>();
+        options.add(HOST);
+        options.add(PORT);
+        options.add(USER);
+        options.add(PASSWORD);
+        return options;
+    }
+
+    // …
+}
+```
+
+Now take a look at the `createDynamicTableSource()` function in the `ImapTableSouceFactory` class.  Recall that previously (in part one) you had created a small helper utility [TableFactoryHelper](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/table/factories/FactoryUtil.TableFactoryHelper.html), that Flink offers which ensures that required options are set and that no unknown options are provided. You can now use it to automatically make sure that the required options of hostname, port number, username, and password are all provided when creating a table using this connector. The helper function will throw an error message if one required option is missing. You can also use it to access the provided options (`getOptions()`), convert them into an instance of the `ImapTableSource` class created earlier, and provide the instance to the table source:
+
+// why would you want to do the latter??
+
+```java
+public class ImapTableSourceFactory implements DynamicTableSourceFactory {
+
+    // ...
+
+    @Override
+    public DynamicTableSource createDynamicTableSource(Context ctx) {
+        final FactoryUtil.TableFactoryHelper factoryHelper = FactoryUtil.createTableFactoryHelper(this, ctx);
+        factoryHelper.validate();
+
+        final ImapSourceOptions options = ImapSourceOptions.builder()
+            .host(factoryHelper.getOptions().get(HOST))
+            .port(factoryHelper.getOptions().get(PORT))
+            .user(factoryHelper.getOptions().get(USER))
+            .password(factoryHelper.getOptions().get(PASSWORD))
+            .build();
+        final List<String> columnNames = ctx.getCatalogTable().getResolvedSchema().getColumnNames();
```

Comment:
actually we should use `toPhysicalDataType` to erase computed columns and time attributes, but the problem in this tutorial is actually that it mixes the concepts of metadata columns and physical columns. so it uses physical columns as a replacement for metadata columns. `(RowType) ctx.getCatalogTable().getResolvedSchema().toPhysicalRowDataType().getLogicalType().getFieldNames()` but this is not very nice for an example :(

Reply:
I discussed this offline with @twalthr. In the interest of finding a balance between "correctness" and brevity, let's change this to this:

```
        final List<String> columnNames = ctx.getCatalogTable().getResolvedSchema().getColumns().stream()
            .filter(Column::isPhysical)
            .map(Column::getName)
            .collect(Collectors.toList());
```

Additionally we should add an explanation that ideally we'd be making use of connector metadata instead (with a link), but that we'll keep it simple here instead and maybe refer again to github.com/TNG/flink-connector-email which does implement this using metadata now.